nutils/function.py

```diff
@@ -465,6 +465,262 @@ def lower(self, *, points_shape: Tuple[evaluable.Array, ...] = (), transform_cha
+
+    Returns
+    -------
+    :class:`Array` or similar
+        The partial derivative of this function to the given argument.
+    '''
+
+    raise NotImplementedError('The partial derivative of {} to argument {} (counting from 0) is not defined.'.format(type(self).__name__, iarg)) # pragma: nocover
+
+class _CustomEvaluable(evaluable.Array):
+
+  def __init__(self, name, evalf, partial_derivative, args: Tuple[Union[evaluable.Array, evaluable.EvaluableConstant], ...], shape: Tuple[int, ...], dtype: DType, points_shape: Tuple[evaluable.Array, ...], transform_chains: Tuple[evaluable.TransformChain, ...], coordinates: Tuple[evaluable.Array, ...]) -> None:
+    assert all(isinstance(arg, (evaluable.Array, evaluable.EvaluableConstant)) for arg in args)
+    self.name = name
+    self.custom_evalf = evalf
+    self.custom_partial_derivative = partial_derivative
+    self.args = args
+    self.points_dim = len(points_shape)
+    self.lower_args = dict(points_shape=points_shape, transform_chains=transform_chains, coordinates=coordinates)
+    super().__init__((evaluable.Tuple(points_shape), *args), shape=points_shape+shape, dtype=dtype)
+
+  @property
+  def _node_details(self) -> str:
+    return self.name
+
+  def evalf(self, points_shape: Tuple[numpy.ndarray, ...], *args: Any) -> numpy.ndarray:
+    points_shape = tuple(n.__index__() for n in points_shape)
+    npoints = util.product(points_shape, 1)
+    # Flatten the points axes of the array arguments and call `custom_evalf`.
+    flattened = (arg.reshape(npoints, *arg.shape[self.points_dim:]) if isinstance(origarg, evaluable.Array) else arg for arg, origarg in zip(args, self.args))
+    result = self.custom_evalf(*flattened)
+    assert result.ndim == self.ndim + 1 - self.points_dim
+    # Unflatten the points axes of the result. If there are no array arguments,
+    # the points axis must have length one. Otherwise the length must be
+    # `npoints` (checked by `reshape`).
+    if not any(isinstance(origarg, evaluable.Array) for origarg in self.args):
+      if result.shape[0] != 1:
+        raise ValueError('Expected a points axis of length one but got {}.'.format(result.shape[0]))
+      return numpy.broadcast_to(result[0], points_shape + result.shape[1:])
+    else:
+      return result.reshape(points_shape + result.shape[1:])
+
+  def _derivative(self, var: evaluable.Array, seen: Dict[evaluable.Array, evaluable.Array]) -> evaluable.Array:
+    if self.dtype != float:
+      return super()._derivative(var, seen)
+    result = evaluable.Zeros(self.shape + var.shape, dtype=self.dtype)
+    unlowered_args = tuple(_Unlower(arg, **self.lower_args) if isinstance(arg, evaluable.Array) else arg.value for arg in self.args)
+    for iarg, arg in enumerate(self.args):
+      if not isinstance(arg, evaluable.Array) or arg.dtype != float or var not in arg.dependencies and var != arg:
+        continue
+      d1 = self.custom_partial_derivative(iarg, *unlowered_args).lower(**self.lower_args)
```

Comment:
I expected the following shape check to work but for some reason the unit tests are failing:
```
if d1.shape != self.shape[self.points_dim:] + arg.shape:
        raise Exception('invalid partial derivative to argument #{}'.format(iarg))
```

Reply:
I've rewritten the check in terms of `function.Array` instead of `evaluable.Array` and extended the error message with the expected and received shape.